Explainer on user-defined operators

cameel opened the pull request:
Article adapted from https://github.com/ethereum/solidity/issues/13718#issuecomment-1341058649.

Initial version, still needs some tweaks and perhaps a bit more in terms of introduction and closing thoughts.

NunoFilipeSantos commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,173 @@
+category: Explainers
+---
+
+The prominent new feature in [Solidity 0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18)
+is the support for defining operators on [user defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs).
+Let's take a closer look into the details of the the new mechanism.
```
```suggestion
The prominent new feature in [Solidity 0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18)
is the support for defining operators on [user defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs).
A UDVT is a type of abstraction that allows creating a zero-cost representation of an elementary value type. 
This is similar to creating an alias, but with stricter type requirements.

Bellow, we're going to take a closer look into the details of this new mechanism.
```

NunoFilipeSantos commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,173 @@
+| Ternary             | `?:`                                                                    |
+| Other               | `delete`, `new`                                                         |
+
+Most of these operators work only with [value types](https://docs.soliditylang.org/en/latest/types.html#value-types).
+
+In particular, with the exception of `?:` and `=`, these operators are not available on user-defined
+value types, making their use much less convenient than the built-in ones.
+One way to make the use of these types more convenient is attaching helper functions with
+[`using for`](https://docs.soliditylang.org/en/latest/contracts.html#using-for).
```
```suggestion
Most of these operators work only with [value types](https://docs.soliditylang.org/en/latest/types.html#value-types).
With the exception of `?:` and `=`, all of these operators are not available on user-defined
value types, making their use much less convenient than the built-in ones.
One way to make the use of these types more convenient is attaching helper functions with
[`using for`](https://docs.soliditylang.org/en/latest/contracts.html#using-for).
```

NunoFilipeSantos commented on _posts/2023-02-01-user-defined-operators.md:
````diff
@@ -0,0 +1,173 @@
+One way to make the use of these types more convenient is attaching helper functions with
+[`using for`](https://docs.soliditylang.org/en/latest/contracts.html#using-for).
+
+### Quick Refresher on `using for`
+`using for` can attach a function to a type.
+
+```solidity
+type Int is int;
+using {add} for Int;
+
+function add(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function test(Int a, Int b) pure returns (Int) {
+    return a.add(b);
+}
+```
+
+Currently it has 3 forms:
+1. `using L for T`: attaches all non-private functions from library `L` to type `T`.
+1. `using L for *`: attaches all non-private functions from library `L` to any type, including types that
+    cannot be named in the language (e.g. types of literals or array slices).
+1. `using {f, g, h} for T`: Attaches functions `f`, `g` and `h` to type `T`.
+
+`using for` can be used at file level or inside contracts, including libraries (but not interfaces).
+
+At file level the directive can be marked as `global` which extends its effect to all source units
+where a type is visible.
+The use of `global` is possible only with UDVTs, structs and enums that are defined in the current source unit.
+
+The following example shows off the main features of `using for` in combination wth UDVTs:
+
````
```suggestion
### Quick Refresher on `using for`
The directive `using A for B;` can be used to attach functions (A) as member functions to any type (B). 
These functions will receive the object they are called on as their first parameter.

```solidity
type Int is int;
using {add} for Int;

function add(Int a, Int b) pure returns (Int) {
    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
}

function test(Int a, Int b) pure returns (Int) {
    return a.add(b);
}
```

Currently, it has 3 forms:
1. `using L for T`: attaches all non-private functions from library `L` to type `T`.
1. `using L for *`: attaches all non-private functions from library `L` to any type, including types that
    cannot be named in the language (e.g. types of literals or array slices).
1. `using {f, g, h} for T`: Attaches functions `f`, `g`, and `h` to type `T`.

✅ Can be used:
- at the file level
- inside contracts,  including libraries

❌ Can not be used
- Inside contracts with interfaces

At the file level, the directive can be marked as `global`, which extends its effect to all source units
where a type is visible.
The use of `global` is possible only with UDVTs, structs, and enums that are defined in the current source unit.

The following example shows off the main features of `using for` in combination with UDVTs:

```

NunoFilipeSantos commented on _posts/2023-02-01-user-defined-operators.md:
````diff
@@ -0,0 +1,173 @@
+using {add as +} for Int;
+
+function add(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function test(Int a, Int b) pure returns (Int) {
+    return a + b;
+}
+```
+
+Note that defining an operator for a type does not automatically make it available as an attached function.
+E.g. `using {add as +} for Int` does not imply `using {add} for Int`.
+
+#### Which Operators Can Be User-Defined?
+| Category   | Operators                        |
+|------------|----------------------------------|
+| Bitwise    | `&`, `\|`, `^`,  `~`             |
+| Arithmetic | `+`, `-`, `*`, `/`, `%`          |
+| Comparison | `==`, `!=`, `<`, `<=`, `>`, `>=` |
+
+Of the operators listed above `!` and `~` are unary, `-` can be both unary and binary and the
+rest is always binary.
+Note that Solidity does not have binary `+`.
+
+Several operators have been excluded from the initial implementation for the sake of simplicity.
+We are considering enabling them in one of the future releases if the design issues can be
+resolved satisfactorily:
+1. `&&`, `||` and `!` are not necessary for most UDVTs because user-defined comparison operators
+    must return `bool`, for which the built-in boolean operators are sufficient.
+
+    In case of `&&` and `||`  an additional complication is their special short-circuiting behavior.
+1. `<<`, `>>` and `**` have non-uniform arguments.
+    For built-in types the right-hand-side is always an unsigned type.
+1. `=` is one of the few operators that already work with UDVTs.
+    The new feature is not meant to allow redefining existing operators.
+1. Assignment and increment/decrement operators in general modify their argument and return a reference.
+
+    `++` and `--` also have separate prefix and postfix variants with different semantics.
+
````
```suggestion
### Introducing User-Defined Operators
The new feature adds another form of the `using for` syntax, where instead of attaching a function,
it is assigned an operator symbol.
The function is then invoked whenever the operator is used on the values of the type:

```solidity
type Int is int;
using {add as +} for Int;

function add(Int a, Int b) pure returns (Int) {
    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
}

function test(Int a, Int b) pure returns (Int) {
    return a + b;
}
```

Note that defining an operator for a type does not automatically make it available as an attached function.
E.g. `using {add as +} for Int` does not imply `using {add} for Int`.

#### Which Operators Can Be User-Defined?
| Category   | Operators                        |
|------------|----------------------------------|
| Bitwise    | `&`, `\|`, `^`,  `~`             |
| Arithmetic | `+`, `-`, `*`, `/`, `%`          |
| Comparison | `==`, `!=`, `<`, `<=`, `>`, `>=` |

Of the operators listed above, `!` and `~` are unary, `-` can be both unary and binary and the
rest is always binary.
Note that Solidity does not have binary `+`.

Several operators have been excluded from the initial implementation for the sake of simplicity.
We are considering enabling them in one of the future releases if the design issues can be
resolved satisfactorily:
1. `&&`, `||` and `!` are not necessary for most UDVTs because user-defined comparison operators
    must return `bool`, for which the built-in boolean operators are sufficient.
    In case of `&&` and `||`  an additional complication is their particular short-circuiting behavior.
1. `<<`, `>>` and `**` have non-uniform arguments.
    For built-in types, the right-hand side is always an unsigned type.
1. `=` is one of the few operators that already work with UDVTs.
    The new feature is not meant to allow redefining existing operators.
1. Assignment and increment/decrement operators, in general, modify their argument and return a reference.

    `++` and `--` also have a separate prefix and postfix variants with different semantics.

```

NunoFilipeSantos commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,173 @@
+
+    `++` and `--` also have separate prefix and postfix variants with different semantics.
+
+#### Which Types Can Have User-Defined Operators?
+Currently operators can only be defined for user-defined value types.
+We are also considering enabling them for structs and enums in future releases.
+There are no plans to allow defining operators for built-in value types, especially in cases where
+an operator is already provided by the compiler.
+
+Note that the type wildcard (`*`) is not allowed in `using for` when defining operators, which
+is in line with how it already worked when used attach individual functions.
+
+Since currently there are no implicit conversions allowed to and from UDVTs, user-defined operators
+must always be invoked with exact types.
+
```
```suggestion
#### Which Types Can Have User-Defined Operators?
Currently, operators can only be defined for user-defined value types.
We are also considering enabling them for structs and enums in future releases.
There are no plans, for now, to allow defining operators for built-in value types, especially in cases where
an operator is already provided by the compiler.

Note that the type wildcard (`*`) is not allowed in `using for` when defining operators, which
is in line with how it already worked when used to attach individual functions.

Since currently there are no implicit conversions allowed to and from UDVTs, user-defined operators
must always be invoked with exact types.

```

NunoFilipeSantos commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,173 @@
+it is only allowed at file level, the type must be defined in the current source unit and the
+effects extend to wherever the type is visible, not just the current source unit.
+
+#### Can Operators Be Redefined?
+It is not possible to redefine any built-in operator.
+All valid cases are the ones where previously there was no built-in operator.
+
+Only one definition is allowed for each user-defined operators for a given type.
+Note that:
+- Unary and binary `-` count as two separate operators.
+- As long as the definition stays the same, it is not an error to repeat the `using for` directive,
+    just as it is not an error to attach the same function to a type twice.
+
+`using for` is not inherited by derived contracts and this applies to operators as well.
+For this reason, it is possible for a base contract and a derived contract to use different definitions
+for the same operator.
+The same is not possible at file level.
+If an operator is defined at file level, it cannot have a different definition inside a contract where
+that file-level definition is visible.
+
```
```suggestion
#### Can Operators Be Redefined?
It is not possible to redefine any built-in operator.
All valid cases are the ones where previously there was no built-in operator.

As such, only one definition is allowed for each user-defined operator for a given type.
Note that:
- Unary and binary `-` count as two separate operators.
- As long as the definition stays the same, it is not an error to repeat the `using for` directive,
    just as it is not an error to attach the same function to a type twice.

`using for` is not inherited by derived contracts, and this applies to operators as well.
For this reason, it is possible for a base contract and a derived contract to use different definitions
for the same operator.
The same is not possible at the file level.
If an operator is defined at the file level, it cannot have a different definition inside a contract where
that file-level definition is visible.

```

NunoFilipeSantos commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,173 @@
+If an operator is defined at file level, it cannot have a different definition inside a contract where
+that file-level definition is visible.
+
+### AST Changes
+The new feature introduces two new attributes to the AST:
+1. Functions listed in `functionList` on `UsingForDirective` nodes, may now optionally have an
+    `operator` attribute.
+    When present, it indicates that the function defines an operator and specifies its name.
+1. `BinaryOperation` and `UnaryOperation` may now have a `function` attribute.
+    If present, the attribute indicates that the operation is actually an invocation of a user-defined
+    operator and specifies the ID of the function used to define it.
+
+This effectively introduces a new way to perform a function call, not represented by a `FunctionCall` node.
+This will affect any tool that needs to detect and follow function calls, for example to build a
+control flow graph.
```
```suggestion
### AST Changes
The new feature introduces two new attributes to the Abstract Syntax Tree (AST):
1. Functions listed in `functionList` on `UsingForDirective` nodes, may now optionally have an
    `operator` attribute.
    When present, it indicates that the function defines an operator and specifies its name.
1. `BinaryOperation` and `UnaryOperation` may now have a `function` attribute.
    If present, the attribute indicates that the operation is actually an invocation of a user-defined
    operator and specifies the ID of the function used to define it.

This effectively introduces a new way to perform a function call, not represented by a `FunctionCall` node.
This will affect any tool that needs to detect and follow function calls, for example, to build a
control flow graph.
```

NunoFilipeSantos reviewed (commented):
Made a few suggestions.

matheusaaguiar commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,237 @@
+
+Most of these operators work only with [value types](https://docs.soliditylang.org/en/latest/types.html#value-types).
+
+With the exception of `?:` and `=`, all of these operators are not available on user-defined
+value types, making their use much less convenient than the built-in ones.
```
```suggestion
With the exception of `?:` and `=`, all of these operators are not available on UDTVs, 
making their use much less convenient than the built-in ones.
```

cameel replied:
I'm using them interchangeably for some variety so it was intentional.

matheusaaguiar commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,237 @@
+The use of `global` is possible only with UDVTs, structs, and enums that are defined in the current source unit.
+
+### Introducing User-Defined Operators
+The new feature adds another form of the `using for` syntax, where instead of attaching a function,
+it is assigned an operator symbol.
```
```suggestion
The new feature in [Solidity 0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18)
adds another form of the `using for` syntax, where a function can be assigned to an operator symbol.
```

matheusaaguiar commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,237 @@
+| Comparison | `==`, `!=`, `<`, `<=`, `>`, `>=` |
+
+Of the operators listed above, `~` is unary, `-` can be both unary and binary and the rest is always binary.
+Note that Solidity does not have binary `+`.
```
```suggestion
Note that Solidity does not have unary `+`.
```

cameel replied:
Nice catch!

matheusaaguiar commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,237 @@
+must always be invoked with exact types.
+
+#### What Are the Restrictions on User-Defined Operator Definitions?
+The operator definition must be a pure function, defined at the file level or a library.
```
```suggestion
The operator definition must be a pure function, defined at the file level or inside a library.
```

matheusaaguiar commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,237 @@
+
+#### What Are the Restrictions on User-Defined Operator Definitions?
+The operator definition must be a pure function, defined at the file level or a library.
+All parameters must be of the same type, which must be the type specified in `using for`.
```
```suggestion
All parameters must be of the same type specified in `using for`.
```

cameel replied:
I'll try to reword it, but the proposed version sounds a bit ambiguous to me. As if the parameters were defined in `using for` :)

matheusaaguiar commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,237 @@
+- Overloaded functions are not purposefully disallowed, but due to the lack of a language mechanism
+    for selecting a specific overload, it is currently not possible to use such functions.
+- Built-in functions like `keccak256`, `revert` or `abi.encode()` cannot be used even if their
+    parameter and return value types happen to match the type.
```
```suggestion
    parameter and return value types happen to match the type specified in `using for`.
```

matheusaaguiar commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,237 @@
+- Built-in functions like `keccak256`, `revert` or `abi.encode()` cannot be used even if their
+    parameter and return value types happen to match the type.
+
+The rules for `using for global` with operators are the same as for bound functions:
```
```suggestion
The rules for `using for global` with operators are the same as for attached functions:
```

matheusaaguiar commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,237 @@
+1. `=` is one of the few operators that already worked with UDVTs.
+    The new feature is not meant to allow redefining existing operators so providing a custom
+    definition for it will not be possible.
+1. Other assignment operators modify their argument and return a reference.
```
```suggestion
1. Other assignment operators, as `+=`, for example, modify their argument and return a reference.
```

matheusaaguiar commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,237 @@
+1. `++` and `--` modify their argument and also have a separate prefix and postfix variants with
+    different semantics.
+
+We are considering enabling them in one of the future releases, provided that the design issues can
+be resolved satisfactorily.
```
```suggestion
We are considering enabling them, (except `&&`, `||` and `!`, as mentioned before )
in one of the future releases, provided that the design issues can be resolved satisfactorily.
```

cameel replied:
I mean, that includes also also `=`. I wanted the list above to indicate which those are instead of repeating them. I'll see if I can reword it.

matheusaaguiar reviewed (commented):
I will work on adding more text and examples later.

NunoFilipeSantos commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,237 @@
+The main thing holding us back are the planned changes to reference types, in particular the fact
+that data location (`storage`, `memory`, etc.) is going to become a part of the type itself.
+To avoid having to introduce breaking changes later, we decided to defer support for structs until
+these changes are behind us.
```
```suggestion
these changes are behind us.

#### Why should you use this
@matheusaaguiar ?
```

matheusaaguiar replied:
Working on this.

matheusaaguiar commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,237 @@
+category: Explainers
+---
+
+The prominent new feature in [Solidity 0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18)
```
```suggestion
The prominent new feature in Solidity 0.8.19
```

cameel commented on _posts/2023-02-01-user-defined-operators.md:
```diff
@@ -0,0 +1,237 @@
+---
+
+The prominent new feature in [Solidity 0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18)
+is the support for defining operators on [user defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs).
```
Added an emphasis. I think bolding it would be a bit too strong.

cameel commented:
Added requested corrections, updated dates, versions, etc.

Now working on updating the text to match the changes I made to the PR since I wrote this.

NunoFilipeSantos commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,332 @@
+must always be invoked with exact types.
+
+#### What Are the Restrictions on User-Defined Operator Definitions?
+The operator definition must be a pure function, defined at the file level or inside a library.
+The type specified in `using for` must be the type of all parameters and the return value.
+The only exception are comparison operators, where the return type must be `bool`.
+
+A few points of note:
+- Being a `pure` function, the operator definition is allowed to make `pure` external calls.
+- Library functions are allowed but contract and interface functions are not.
+    This means that the definition cannot be virtual.
+- Private library functions are allowed but visible only within the library.
+- Operator definition may be a public library function.
+    External functions will eventually be allowed too
+    (see [issue #13765](https://github.com/ethereum/solidity/issues/13765)).
+- Overloaded functions are not purposefully disallowed, but due to the lack of a language mechanism
+    for selecting a specific overload, it is currently not possible to use such functions.
+- Built-in functions like `keccak256`, `revert` or `abi.encode()` cannot be used even if their
+    parameter and return value types happen to match the type specified in `using for`.
+
+The rules for `using for global` with operators are the same as for attached functions:
+it is only allowed at the file level, and the type must be defined in the current source unit and the
+effects extend to wherever the type is visible, not just the current source unit.
+
+In the next example, library `L` is used to group functions assigned to arithmetic operators for type `Int`.
+We also add more operators, with functions `unsub` and `equals` being respectively assigned to unary `-` and comparison `=`.
```
```suggestion
The operator definition must be a pure function, defined at the file level or inside a library.
The type specified in `using for` must be the type of all parameters and the return value.
The only exception are comparison operators, where the return type must be `bool`.

A few points of note:
- Being a `pure` function, the operator definition is allowed to make `pure` external calls.
- Library functions are allowed but contract and interface functions are not.
    This means that the definition cannot be virtual.
- Private library functions are allowed but visible only within the library.
- Operator definition may be a public library function.
    External functions will eventually be allowed too
    (see [issue #13765](https://github.com/ethereum/solidity/issues/13765)).
- Overloaded functions are not purposefully disallowed, but due to the lack of a language mechanism
    for selecting a specific overload, it is currently not possible to use such functions.
- Built-in functions like `keccak256`, `revert` or `abi.encode()` cannot be used even if their
    parameter and return value types happen to match the type specified in `using for`.

The rules for `using for global` with operators are the same as for the attached functions:
it is only allowed at the file level, and the type must be defined in the current source unit and the
effects extend to wherever the type is visible, not just the current source unit.

In the next example, library `L` is used to group functions assigned to arithmetic operators for type `Int`.
We also add more operators, with functions `unsub` and `equals` being respectively assigned to unary `-` and comparison `=`.
```

NunoFilipeSantos replied:
This correction only applies if you're mentioning the functions here, if it's a generic thing, disregard it.

cameel replied:
Yeah, I meant "attached functions" as a general concept, not some specific functions.

NunoFilipeSantos commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,332 @@
+very likely to become available sooner or later.
+
+#### Operators on Structs
+While UDVTs were the primary target, not all types can reasonably fit into a single slot.
+The way to solve this problem is to allow operators taking and returning structs.
+The main thing holding us back are the planned changes to reference types, in particular the fact
+that data location (`storage`, `memory`, etc.) is going to become a part of the type itself.
+To avoid having to introduce breaking changes later, we decided to defer support for structs until
+these changes are behind us.
```
```suggestion
While UDVTs were the primary target, not all types can reasonably fit into a single slot.
The way to solve this problem is to allow operators to take and return structs.
The main thing holding us back are the planned changes to reference types, in particular the fact
that data location (`storage`, `memory`, etc.) is going to become a part of the type itself.
To avoid having to introduce breaking changes later, we decided to defer support for structs until
these changes are behind us.
```

NunoFilipeSantos commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,332 @@
+To avoid having to introduce breaking changes later, we decided to defer support for structs until
+these changes are behind us.
+
+#### Why should you use this
+There are some advantages in defining operators for UDVTs. It allows for more expressive code where the user
+can define the exact behaviour of an operation for a desired type. Besides that, it improves maintainability
+and readability of the code. It is possible to combine several operations in a single clear expression.
+Additionally, the use of custom defined operators also provides a type-safe interface for the desired type.
```
```suggestion
#### Why should you use this
There are some advantages to defining operators for UDVTs. It allows for more expressive code where the user
can define the exact behavior of an operation for a desired type. 
Besides that, it improves the maintainability and readability of the code. It is possible to combine several operations in a single clear expression.
Additionally, the use of custom-defined operators also provides a type-safe interface for the desired type.
```

NunoFilipeSantos reviewed (dismissed):
A few tweaks are suggested.

matheusaaguiar commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,362 @@
+The use of `global` is possible only with UDVTs, structs, and enums that are defined in the current source unit.
+
+### Introducing User-Defined Operators
+The new feature extends the third form of the `using for` syntax, so that the function, instead of being
+attached to the type, it is instead assigned an operator symbol.
```
```suggestion
The new feature extends the third form of the `using for` syntax, so that the
function is assigned an operator symbol, instead of being attached to the type.
```

matheusaaguiar commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,362 @@
+effects extend to wherever the type is visible, not just the current source unit.
+
+In the next example, library `L` is used to group functions assigned to arithmetic operators for type `Int`.
+We also add more operators, with functions `unsub` and `equals` being respectively assigned to unary `-` and comparison `=`.
```
```suggestion
We also add more operators, with functions `unsub` and `equals` being respectively assigned to unary `-` and comparison `==`.
```

ekpyron commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+ To avoid having to introduce breaking changes later, we decided to defer support for structs until
+these changes are behind us.
+
+We want to eventually support operators on other reference types, but it is likely that this usage will
+require naming them, using some kind of yet-to-be-introduced user-defined reference type feature similar
+to UDVTs, or at least defining a type alias.
```
This is a bit much conjecturing... it's equally possible that we allow user-defined operators on structs themselves, so I'd just skip this part.

cameel replied:
Removed.

ekpyron commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+to UDVTs, or at least defining a type alias.
+
+#### Operators on Mixed Types
+Inability to define operators with arguments of different types is somewhat constraining and lifting this
+limitation is also a part of our vision for UDVTs.
+There are, however, multiple ways to achieve this, and explicitly defining an operator for every type combination
+is only one of them.
+Other possibilities involve generics specialized with traits or having implicit conversions.
+
+This part will unfortunately have to wait until we are done with the changes to the type system that are on our current roadmap.
+Doing them before would only result in something that would limit our design choices, and would have to be discarded
+in a breaking release.
```
I'm not sure this is true :-). Mixed type operators are a significant problem for type inference, so I'm not sure we'll ever want to have them.

I think it would be better to be very conservative on this and merely state that the first version we're releasing is a minimal viable implementation that will definitely not cause problems if considered as part of a generic type system or trait system with type inference without making any promises or talking about mixed operators as part of a vision, which they aren't necessarily.

cameel replied:
How about this?

```
Inability to apply operators to values of different types is somewhat constraining and lifting this
limitation is something we would like to address eventually, though it is not yet clear in what form.
There are multiple ways to achieve this.
Explicitly defining an operator for every type combination is just one of them, and not necessarily the most elegant.
For the time being, we can only say that the feature we are releasing is a minimal viable implementation
that will definitely not cause problems if considered as part of a generic type system or trait system
with type inference.
```

But I'm not sure I understand what you mean. Are you saying that ending up without operators that work on more than one type is an real possibility? I thought we wanted to allow that, one way or another. I.e. the classic example of adding `Timestamp` and `Duration` that @chriseth often brings up. That has to be possible eventually, doesn't it?

cameel replied:
The whole section about the future of operators removed.

ekpyron replied:
I swayed @chriseth away from it by now :-). Please don't promise mixed operators in any way, we'll likely never have them and it will be hard to explain to people why already, since it's a type-theoretic problem of the ambiguity of type inference that's at stake here that's not so easy to convey.

nikola-matic commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+
+### A Quick Refresher
+Before we get into the details of the new feature and how it affects the language, it may be a good
+idea to first go over the two existing features that it builds upon: built-in operators and `using for`.
```
Can we link to the appropriate place in the docs for built-in-operators and `using-for`?

cameel replied:
They are already linked, at the beginning of each respective section.

nikola-matic commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+Most of them require no explanation.
+In some cases, however, you may be surprised by what counts as an operator.
+The common infix operator, like `a + b`, that may be on your mind when you hear "operator" is
+not the only flavor operators come in.
```
```suggestion
not the only flavour operators come in.
```
Are we using British English spelling, or have we given up on that?

ekpyron replied:
Tell that to our optimiser :-).

nikola-matic replied:
> Tell that to our optimiser :-).

This [one](https://docs.soliditylang.org/en/latest/internals/optimizer.html)?

ekpyron replied:
This one: https://github.com/ethereum/solidity/tree/develop/libyul/optimiser
Or this one: https://github.com/ethereum/solidity/blob/develop/libevmasm/ConstantOptimiser.h
But don't start testing them (https://github.com/ethereum/solidity/blob/develop/test/libyul/YulOptimizerTest.h)...

ekpyron replied:
This one is the best: https://github.com/ethereum/solidity/blob/develop/libyul/optimiser/OptimizerUtilities.h

matheusaaguiar replied:
> Is @matheusaaguiar gonna expand on the examples on this branch, or as part of another PR?

Added it in https://github.com/ethereum/solidity-blog/pull/202/commits/c50767711a9cfed23a8a7fb1595f5300900caa03

cameel replied:
I'm intentionally using the US spelling here because it's just a blog post I'm writing, not a part of our official docs, and it's just much easier for me to write that way. If I don't pay close attention, I automatically revert to that spelling by the sheer force of habit :)

> This one is the best: https://github.com/ethereum/solidity/blob/develop/libyul/optimiser/OptimizerUtilities.h

Well, it's just very inconsistent. It was very frustrating when I first worked with the optimizer and tried to stick to British spelling :) I eventually just gave up because it was already beyond repair.

ekpyron commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+    We are leaning towards the latter, but it needs more consideration.
+1. `++` and `--` are very convenient, and we want to have them eventually.
+    They just need to be handled in a special way, and there is more than way to do so.
+1. Composite assignment operators, such as `+=`, will likely be provided automatically
+    in the future as long as the corresponding arithmetic/bitwise operator is available.
+    Allowing a direct implementation is currently not impossible anyway, because there are no references
+    to value types in the language.
```
I guess that was supposed to be "not possible" instead of "not impossible" :-)? Also generally a lot of conjecture again, we may also still decide to let them be freely implemented - but fine with keeping it here.

cameel replied:
> I guess that was supposed to be "not possible" instead of "not impossible" :-)?

Yes :)

cameel replied:
I know it's not certain, that's why it's not worded more strongly. It's not like we're promising this, it's just an overview of what *might* happen as far as we can tell today.

ekpyron commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+is the support for defining *operators* on [user defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs).
+
+If you have not been keeping up with recent features, UDVTs are a new class of types introduced in Solidity 0.8.8.
+They provide zero-cost abstraction of an elementary value type that results in a completely new type.
```
Personally, I'm not a fan of the weird rusticean "zero-cost abstraction" thing :-). Fine with keeping it, but I'd just say "an abstraction". But also fine with keeping it.

cameel replied:
That was actually requested by @NunoFilipeSantos in https://github.com/ethereum/solidity-blog/pull/202#discussion_r1085549651 so I just went with it and just adapted it. For me either way is fine. Changed to this now:
```
They provide an abstraction over an elementary value type that results in a completely new type.
```

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+---
+
+The prominent new feature in [Solidity 0.8.19](https://github.com/ethereum/solidity/releases/tag/v0.8.19)
+is the support for defining *operators* on [user defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs).
```
```suggestion
is the support for defining *operators* on [user-defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs).
```

nikola-matic reviewed (commented):
Is @matheusaaguiar gonna expand on the examples on this branch, or as part of another PR?

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+
+If you have not been keeping up with recent features, UDVTs are a new class of types introduced in Solidity 0.8.8.
+They provide zero-cost abstraction of an elementary value type that results in a completely new type.
+This is similar to creating an alias, but the new type is distinct from the underlying value type
```
I'm not sure I would say that it is similar to creating an alias, because it is not ;)
Also what do you mean by "all other UDVTs derive from it"?

cameel replied:
> I'm not sure I would say that it is similar to creating an alias, because it is not ;)

This is another thing that was suggested to me: https://github.com/ethereum/solidity-blog/pull/202#discussion_r1085549651. I mean, I know it's not an alias, which is why I reworded the original suggestion, but still, it's not like there are similarities at all. That's all it says and the later part of the sequence points out what's not similar.

> Also what do you mean by "all other UDVTs derive from it"?

All the UDVTs that use the same underlying built-in type.

---

Anyway, I'm fine with rewording it, or even removing that part of the introduction (in my original version it wasn't there anyway). What would be your suggestion for the new wording?

chriseth replied:
Oh, I thought "it" meant the new type...

cameel replied:
Changed to `derived from that underlying type`.

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+#### Built-in operators
+Solidity, like the vast majority of programming languages,
+[offers a set of operators](https://docs.soliditylang.org/en/latest/types.html#operators) as syntactic sugar
+over common mathematical functions.
```
```suggestion
over common arithmetical and logical functions.
```

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+[offers a set of operators](https://docs.soliditylang.org/en/latest/types.html#operators) as syntactic sugar
+over common mathematical functions.
+The difference between operators and functions is purely syntactical, and the decision
+to choose one over the other is to a large extent arbitrary.
```
```suggestion
to choose one over the other is to a large extent a matter of readability.
```

NunoFilipeSantos replied:
@chriseth I would change to indicate which one is more readable (if that's the case).

cameel replied:
@chriseth I applied the suggestion, but actually, I'm not sure I 100% agree. I think that often the decision *is* to a large degree arbitrary, or at least not always based on hard, absolute criteria.

@NunoFilipeSantos
> I would change it to indicate which one is more readable (if that's the case).

Not sure how to fit it well into the text but I tried. How about this?

```
The use of operators makes expressions more concise, which is helpful as long as they are well
understood by the reader.
```

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+The common infix operator, like `a + b`, that may be on your mind when you hear "operator" is
+not the only flavor operators come in.
+The last row of the table has items that to some users may not seem operator-like at all.
+`delete` and `new`  are in fact operators, and follow precedence and associativity rules despite not being symbols.
```
Do we need to go into so much detail here?

cameel replied:
The whole section is kinda obvious if you know Solidity well, and is meant mostly for people who are not that familiar with Solidity (or even any other language). I thought it could be interesting to that part of the audience.

Also, I think that showing how weird these operators are makes it clearer why the set of definable operators is so limited, and why we did not decide to allow redefining every single operator.

But it's not essential. I'll remove it.

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+- ❌ Inside interface definitions.
+
+When used inside contracts, it is not inherited.
+Its effects extend only to the base contract that has it and not to any other contracts derived from it.
```
"base contract" might be confusing here talking about inheritance.

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+
+At the file level, the directive affects all the free functions, contracts and libraries defined within that file.
+It does not affect other files that import the file containing it, unless it is marked as `global`.
+`global` extends its effect to all files where the type is visible.
```
Also to files where it is invisible. It is usable on all expressions of that type, it does not matter that the type is visible via name.

chriseth replied:
I mean this is the same with contracts already: Functions on a contract can be called on all expressions of that contract type, you do not need the type of the contract visible.

cameel replied:
Very good point. I did not think of that. Changed to just `all files`.

ekpyron commented on _posts/2023-02-22-user-defined-operators.md:
````diff
@@ -0,0 +1,484 @@
+}
+```
+
+An operator can only be bound in a `global` directive and the definition must be a `pure` free function.
````
Here it would actually be nice to justify this a bit more. I.e. the primary use case for this is type libraries that build up a complete type that should always work the same everywhere and we don't have meanings of operators to depend on more than the operator types (i.e. not on scopes, etc.) to avoid pitfalls. Something along those lines :-).

cameel replied:
Yeah, I go into that more in the section on auditing. I generally tried to avoid too much detail in the intro section and initially had that in more specific sections instead. Now those sections are mostly gone though, as the one on auditing turned out to be a good place to list all the important corner cases in a natural way.

ekpyron commented on _posts/2023-02-22-user-defined-operators.md:
````diff
@@ -0,0 +1,484 @@
+```
+
+An operator can only be bound in a `global` directive and the definition must be a `pure` free function.
+The type specified in `using for` must be a user-defined value type and must be the type
+of all the parameters of the function and its return value.
+The only exception are comparison operators, where the return type must be `bool`.
````
Similar to the docs in the actual PR, I'd prefer to say that each operator has a unique signature and to list them... but if you think that table is too much for here, we can keep it like this.

cameel replied:
I think it's fine in the context of a blog post. It's an accurate summary of the current situation and the post is written from the perspective of this point in time, no one will expect it to be updated with the language.

And yeah, I think including the whole table would make it too long (the post is already very long). I really want this bit to be concise.

ekpyron replied:
My problem is that to me the phrasing suggests a conceptually wrong view on the operators :-). There is only one global comparison operator. It's always the same. It always returns ``bool``. You just implement it for types.
The current phrasing suggests that you have multiple unrelated operators flying around and then impose restrictions on their types.
Of course, in the end observationally the result is the same, but it's quite different conceptualizations.

ekpyron replied:
But yeah, as I said originally, we can keep it, if need be :-).

cameel replied:
> The current phrasing suggests that you have multiple unrelated operators flying around and then impose restrictions on their types.
Of course, in the end observationally the result is the same, but it's quite different conceptualizations.

Fair enough. What if I say `comparison operator definitions`? Because that's what I really meant. I probably should have been more precise about this. Currently you *can* have "multiple unrelated operator *definitions* flying around".

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
````diff
@@ -0,0 +1,484 @@
+```
+
+An operator can only be bound in a `global` directive and the definition must be a `pure` free function.
+The type specified in `using for` must be a user-defined value type and must be the type
````
```suggestion
The type specified in `using for` must be a user-defined value type defined in the same file (as is always the case when using `global`) and must be the type
```

cameel replied:
I intentionally did not go into the consequences of making it global here. I wanted this bit to be concise and go to into more details in later sections. Currently the section on auditing lists corner cases and explains the consequences of this a bit more.

ekpyron commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+Note that Solidity does not have unary `+`.
+Unary and binary `-` are considered different operators and need separate definitions.
+
+It is not possible to redefine built-in operators.
```
```suggestion
It is not possible to redefine the operators on built-in types.
```

cameel replied:
No, I do mean built-in operators here. For example `=` or `()` do work on UDVTs but you cannot redefine them.

ekpyron replied:
This is still very misleading :-). `+` is a builtin operator... I mean, it's true that you can't redefine it, but you're thinking of it like that's what *is* allowed, don't you?

cameel replied:
How about this?
```
It is not possible to replace operator definitions.
This applies both to the built-in operator definitions provided by the compiler and to operators
bound to a function provided by the user.
Once the operator is bound, the definition cannot be changed.
```

ekpyron commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+Unary and binary `-` are considered different operators and need separate definitions.
+
+It is not possible to redefine built-in operators.
+All valid cases are the ones where previously there was no built-in operator.
```
Not sure I understand this, resp. whether this is true - what do you mean here? It's not like we can define operators everywhere where we didn't have an operator before...

cameel replied:
I mean that it is possible to define an operator for a type only if the compiler does not already provide a built-in operator for that case. But I agree that it could be worded better. How about this?

```
An operator can be defined by the user only when there is no built-in operator already provided by the compiler.
```

ekpyron replied:
Yeah, better, but why not be actually precise? Operators can only be defined on user-defined types, not on builtin types, period.

cameel replied:
Because the thing I want to emphasize is that you cannot redefine built-in operators.

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+All valid cases are the ones where previously there was no built-in operator.
+As such, only one definition is allowed for each user-defined operator for a given type.
+As long as the definition stays the same, it is not an error to repeat the `using for` directive,
+just as it is not an error to attach the same function to a type twice.
```
Maybe we should say: Whether `-` refers to the unary or binary variant depends on the number of parameters of the function.

cameel replied:
> Maybe it would be nice to have an example how to bind both unary and binary minus.

Yeah, I actually suggested to @matheusaaguiar to include it in his bigger example but I see it's not there. I'll try to work it into it somehow.

> Maybe we should say: Whether `-` refers to the unary or binary variant depends on the number of parameters of the function.

Good point. Added.

ekpyron commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+It is not possible to redefine built-in operators.
+All valid cases are the ones where previously there was no built-in operator.
+As such, only one definition is allowed for each user-defined operator for a given type.
+As long as the definition stays the same, it is not an error to repeat the `using for` directive,
+just as it is not an error to attach the same function to a type twice.
```
I don't think this is relevant - this is a fluke of our analysis, not a feature.

cameel replied:
ok, I guess I can remove it.

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
````diff
@@ -0,0 +1,484 @@
+}
+```
+
+In the above example, the function `add()` is used *only* as the operator `+`, and not callable as `x.add()`.
````
```suggestion
In the above example, the function `add()` is used *only* as the operator `+`, and not callable as `x.add()` (but it can of course be called like `add(x, y)`).
```

ekpyron commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+`sub()`, on the other hand, can both be used as the operator `-` and called on the type.
+
+The rules for `using for global` with operators are the same as for attached functions (with an additional restriction that operators *must* be global):
+it is only allowed at the file level, and the type must be defined in the current source unit and the
```
``only allowed at file level``

Also what's the "current" source unit? You mean the source unit of the type it applies to, right? Any statement is always located in its current source unit :-).

cameel replied:
ok, good point. Changed to
```
it is only allowed at file level, the type must be defined in the same file and it makes
the operator available everywhere.
```

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+it is only allowed at the file level, and the type must be defined in the current source unit and the
+effects extend to wherever the type is visible, not just the current source unit.
+
+#### More Examples
```
Maybe we should also mention that `using for` needs to be in the same file but the functions that we are attaching can be imported from somewhere else - at least that was a question in the forum.

cameel replied:
Added a mention. This case is also covered by the example added by @matheusaaguiar.

ekpyron commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+    in the future as long as the corresponding arithmetic/bitwise operator is available.
+    Allowing a direct implementation is currently not impossible anyway, because there are no references
+    to value types in the language.
+1. We are considering adding a C++-like
+    ["spaceship operator"](https://en.wikipedia.org/wiki/Three-way_comparison#High-level_languages),
+    i.e. `<=>`.
+    When defined, the compiler could use it to automatically generate all the other comparison operators.
+
+On the other hand, it is unlikely that we will allow redefining some other operators:
+
+1. `&&` and `||` are not necessary for most UDVTs because user-defined comparison operators
+    must return `bool`, for which the built-in boolean operators are sufficient.
+    The additional complication is their short-circuiting behavior that may or may not be
+    appropriate, depending on the semantics of a particular UDVT.
+1. `=` is already provided by the compiler for all types and it is very unlikely that we will ever allow redefining it.
```
This is a *lot* of conjecture again... not sure we'll have a spaceship operators, not sure we *won't* have boolean conjunction and disjunction... does any of this information really help?

Also the last enumeration has two ``1.``s :-).

cameel replied:
Well, it's worded hypothetically. The conjecture is the point. It's just a blog post, we can afford to speculate. It's not like we're making binding promises here.

I think that talking about these things more in the open will help get us more feedback. It puts the ideas out here, people can react to them.

cameel replied:
So, is any of this actually wrong (not just unlikely?). Or is there are more likely alternative. Feel free to rewrite bits that you think are too far out there. Honestly, it's also a way for me to get these ideas out of you and @chriseth :)

ekpyron replied:
Just remove it, there's no value in engaging in hypotheticals here.

ekpyron replied:
And yes, we should have and should talk about this stuff more up front, but let's keep the post clean of hypotheticals.

cameel replied:
The whole section about the future of operators removed.

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+UDVTs in general were meant to fulfill two needs:
+1. More type-safety when performing operations on values of the same underlying value type that express
+    conceptually different things.
+    For example quantities of different tokens could both be stored as `uint` but adding them without
```
```suggestion
    For example quantities of different tokens or with different dimensions could both be stored as `uint` but adding them without
```

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+1. The ability to define new types using value types as the underlying representation.
+    The best example of this are fixed-point types.
+
+    Fixed-point types were originally going to be an integral part of the language, and the
```
```suggestion
    Fixed-point types were originally planned to be an integral part of the language, and the
```

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+#### Will User-Defined Operators Make Auditing Harder?
+In the feedback we received so far one of the most common concerns is the effect of the feature on contract
+auditing and potential to mask malicious code.
+I want to address these concerns here.
```
```suggestion
We want to address these concerns here.
```

ekpyron commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+1. `=` is already provided by the compiler for all types and it is very unlikely that we will ever allow redefining it.
+
+#### Operators on Reference Types
+While UDVTs were the primary target, not all types can reasonably fit into a single slot.
+The way to solve this problem is to allow operators taking and returning structs.
```
Not sure this tracks. Maybe it's easier to say that we're providing a minimal viable implementation for UDVTs first and consider extending it in the future? We'll probably have multi-stack-slot non-reference types and the way to solve the multi-slot thing is probably *not* always memory structs, etc. :-).

cameel replied:
How about this?

```
#### Operators on Complex Types
While UDVTs were the primary target, in some cases it may also be reasonable to build a custom type
with a more complex underlying representation, such as a struct or a dynamic array.
The main thing holding us back here are upcoming changes to the type system that are a significant part
of our roadmap for the near future.
To avoid having to introduce breaking changes later, we decided to defer support for such operators until
these changes are behind us.
```

cameel replied:
The whole section about the future of operators removed.

ekpyron commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+#### What Will Not Change?
+Due to concerns about the safety of the feature, there are some key elements that are very unlikely to ever change regarding operators.
+Operators will likely always be global and defined at file level with free functions.
+In terms of side-effects, they will stay `pure`.
```
That's probably false.

cameel replied:
Oh, interesting. I know it's possible but I thought it's a very remote possibility, to the point that it would be reasonable to say this.

Also, non-pure operators do not seem to be useful enough to be worth considering.

In any case, removed that part.

ekpyron replied:
Once you have *any* operator on types in storage they're no longer pure, that's no exactly exotic :-).

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+    We decided to leave the choice up to library authors and instead focus on providing primitives
+    necessary to build their own fixed-point types that feel like a native type.
+
+#### Will User-Defined Operators Make Auditing Harder?
```
I think we should stress the advantages of operators a bit more, i.e. readability of `x * t + y * z` as opposed to `x.mul(t).add(y.mul(z))`

cameel replied:
Added an example to show off the readability.

What else should I mention?

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
````diff
@@ -0,0 +1,484 @@
+literals that are already available in the language:
+
+```solidity
+type UFixed16x2 is uint16;
````
Maybe we should use at least 128 bits, which is the common example.

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+1. `<<`, `>>` and `**` were excluded due to their non-uniform arguments.
+    For built-in types, the right-hand side is always an unsigned type.
+1.  `!` was excluded because we have to decide whether it should return the same type as the argument
+    of just `bool`.
```
```suggestion
    or just `bool`.
```

ekpyron commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+        This also makes it impossible to define operators for local types that are defined inside contracts or libraries.
+1. Define `!`, `<<`, `>>` and `**`.
+
+We will definitely lift *some* of these restrictions in the future, but we will consider every case carefully.
```
```suggestion
We may *some* of these restrictions in the future, but we will consider every case carefully.
```

:-)

NunoFilipeSantos commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+The common infix operator, like `a + b`, that may be on your mind when you hear "operator" is
+not the only flavor operators come in.
+The last row of the table has items that to some users may not seem operator-like at all.
+`delete` and `new`  are in fact operators, and follow precedence and associativity rules despite not being symbols.
```
```suggestion
`delete` and `new` are in fact operators and follow precedence and associativity rules despite not being symbols.
```

cameel replied:
Removed this paragraph due to this suggestion: https://github.com/ethereum/solidity-blog/pull/202#discussion_r1113390948

NunoFilipeSantos commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+not the only flavor operators come in.
+The last row of the table has items that to some users may not seem operator-like at all.
+`delete` and `new`  are in fact operators, and follow precedence and associativity rules despite not being symbols.
+Member access via `.`, subscripting arrays with `[]`, or making a function call by adding parentheses
+to a function can be thought of as applications of operators as well.
+Even putting something in parentheses creates a valid operator expression.
+
+It is important to note that in many languages assignment is a statement that is handled specially.
+In Solidity, like in the C-like languages it took inspiration from, assignment can be freely used in
+expressions on par with other operators.
+
+In most cases built-in operators are only defined for [value types](https://docs.soliditylang.org/en/latest/types.html#value-types).
+Notable exceptions are the simple assignment and the "miscellaneous" group.
+For example, you can put almost anything in parentheses or branches of the ternary operator, as a part
+of a bigger expression, even if its type cannot be named or even assigned to anything.
```
```suggestion
Member access via `.`, subscripting arrays with `[]`, or making a function call by adding parentheses
to a function can be thought of as applications of operators as well.
Even putting something in parentheses creates a valid operator expression.

It is important to note that in many languages, an assignment is a statement that is handled specially.
In Solidity, like in the C-like languages it took inspiration from, an assignment can be freely used in
expressions on par with other operators.

In most cases, built-in operators are only defined for [value types](https://docs.soliditylang.org/en/latest/types.html#value-types).
Notable exceptions are the simple assignment and the "miscellaneous" group.
For example, you can put almost anything in parentheses or branches of the ternary operator as a part
of a bigger expression, even if its type cannot be named or even assigned to anything.
```

NunoFilipeSantos commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+| Comparison | `==`, `!=`, `<`, `<=`, `>`, `>=` |
+
+Of the operators listed above, `~` is unary, `-` can be both unary and binary and the rest is always binary.
+Note that Solidity does not have unary `+`.
```
```suggestion
Note that Solidity does not have a unary `+`.
```

NunoFilipeSantos commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+just as it is not an error to attach the same function to a type twice.
+
+#### Operator Bindings vs Attached Functions
+User-defined operators are independent from attached functions.
+One can be done without the other, and both can be done at the same time, even in the same `using for` directive.
+
+The following example illustrates this:
```
```suggestion
User-defined operators are independent of attached functions.
One can be done without the other, and both can be done simultaneously, even in the same `using for` directive.

The following example illustrates this:
```

NunoFilipeSantos commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+    If present, the attribute indicates that the operation is actually an invocation of a user-defined
+    operator and specifies the ID of the function used to define it.
+
+This effectively introduces a new way to perform a function call, not represented by a `FunctionCall` node.
+This will affect any tool that needs to detect and follow function calls, for example, to build a
+control flow graph.
```
```suggestion
This effectively introduces a new way to perform a function call not represented by a `FunctionCall` node.
This will affect any tool that needs to detect and follow function calls, such as building a
control flow graph.
```

cameel replied:
It does not sound correct to me. After the change the second sentence does not really connect with the first one, it refers to a different subject. Not sure how to reword it to make this suggestion work - the way I'd do it is really how I did it already.

NunoFilipeSantos commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+    Fixed-point types were originally going to be an integral part of the language, and the
+    partially implemented [`ufixed` and `fixed`](https://docs.soliditylang.org/en/latest/types.html#fixed-point-numbers)
+    types are the vestiges of that old design.
+    While trying to agree on the final design for the feature, however, we realized that no implementation
+    is going to be a trade-off good enough for most situations and cannot become a suitable default choice.
+    We decided to leave the choice up to library authors and instead focus on providing primitives
+    necessary to build their own fixed-point types that feel like a native type.
```
```suggestion
    While trying to agree on the final design for the feature, however, we realized that no implementation
    will be a good enough trade-off for most situations and cannot become a suitable choice by default.
    We decided to leave the choice up to library authors and instead focus on providing primitives
    necessary to build their own fixed-point types that feel like native types.
```

cameel replied:
Changed, except `suitable choice by default`, which is not quite what I meant. I did mean the "default choice", i.e. a fixed point implementation that is the default one in the language.

NunoFilipeSantos commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+auditing and potential to mask malicious code.
+I want to address these concerns here.
+
+First of all, any new syntax increases the complexity of the language and has some potential to make code more obscure
+and harder to understand when abused.
+It cannot be completely avoided and is always a trade-off between all the good and bad ways it can be used.
+We think that improving the ergonomics of UDVTs is important enough that it outweights any of the downsides we considered.
+
+It is important to realize the actual scope of the feature.
+You may be imagining an elaborate system from languages like C++, with complicated rules and many corner cases.
+As implemented now the feature is deliberately very limited.
+We tried to both limit surprises and avoid prematurely adding support for cases that we have not carefully thought out.
```
```suggestion
First of all, any new syntax increases the complexity of the language and has some potential to make code more obscure
and harder to understand when abused.
It cannot be completely avoided and is always a trade-off between all the good and bad ways it can be used.
We think that improving the ergonomics of UDVTs is important enough that it outweighs any of the downsides we considered.

It is essential to realize the actual scope of the feature.
You may imagine an elaborate system from languages like C++, with complicated rules and many corner cases.
As implemented now the feature is deliberately very limited.
We tried to both limit surprises and avoided prematurely adding support for cases that we have not carefully thought out.
```

NunoFilipeSantos commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+As implemented now the feature is deliberately very limited.
+We tried to both limit surprises and avoid prematurely adding support for cases that we have not carefully thought out.
+
+In the current implementation it is not possible to:
+
+1. Have operators on types other than UDVTs.
+    - User-defined operators on [reference types](https://docs.soliditylang.org/en/latest/types.html#reference-types)
+      like structs or arrays are not allowed.
+    - It is not possible to define them on built-in value types, function types or enums.
+    - You cannot define them to unnamed types using the type wildcard (`*`).
+1. Have operators that are not `pure`.
+    Operators cannot modify storage.
+    They can at most make pure external calls.
+
+    The worst that is technically possible with tricks is for them to modify memory (only through inline assembly)
+    or make staticcalls (through interfaces that do not match actual code deployed on chain).
```
```suggestion
In the current implementation, it is not possible to:

1. Have operators on types other than UDVTs.
    - User-defined operators on [reference types](https://docs.soliditylang.org/en/latest/types.html#reference-types)
      like structs or arrays are not allowed.
    - It is not possible to define them on built-in value types, function types or enums.
    - You cannot define them to unnamed types using the type wildcard (`*`).
1. Have operators that are not `pure`.
    Operators cannot modify storage.
    They can, at most, make pure external calls.

    The worst that is technically possible with tricks is for them to modify memory (only through inline assembly)
    or make staticcalls (through interfaces that do not match actual code deployed on-chain).
```

cameel replied:
Is `on-chain` actually correct here? I mean it's deployed on the chain, not using it as an adjective.

In any case, I reworded the sentence to remove the word :)

NunoFilipeSantos commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+
+    The worst that is technically possible with tricks is for them to modify memory (only through inline assembly)
+    or make staticcalls (through interfaces that do not match actual code deployed on chain).
+1. Have operators on mixed types.
+    Parameter and return value types of operators must be the same.
+    Also, no implicit conversions are possible, so it is very simple to determine if a particular operation
+    will invoke a user-define operator or a built-in one (just look at the type).
+1. Redefine operators.
+    You cannot replace any of the built-in operators with a custom definition.
+    For user-defined operators, only one definition can be provided.
+1. Bind operators to contract functions.
+    This also means that operator definitions cannot be inherited, virtual or remain unimplemented.
+1. Bind operators to overloaded functions.
+1. Use built-in functions to define operators.
+    You cannot e.g. bind an operator directly to `keccak256()`.
+1. Change precedence or associativity of an operator.
+1. Define any of the operators with more complex semantics.
+    You cannot define:
+    - `&&` and `||` due to their short-circuiting behavior.
+    - `++` and `--` because they modify their argument and also have separate prefix and postfix variants with different semantics.
+    - `+=`, `-=`, `^=`, `*=`, `/=`, `%=`, `&=`, `\|=`, `^=`, `<<=`, `>>=` because they modify their argument.
+    - `=`, `()`, `?:` and `.` because they are built-in for all types.
+    - `delete`, `new` and `[]` because they are not usable with UDVTs.
+1. Define custom operators that do not exist in the language, like `===` or unary `+`.
+1. Bind operators in bulk or accidentally.
+    The `using L for T` syntax for attaching a whole library to a type does not cover operators.
+    Operators must always be bound explicitly.
+
+In addition to these restrictions, we decided to limit them even more in response to feedback we received before releasing Solidity 0.8.18.
+This means you cannot also:
+1. Use library functions to define operators.
+    This excludes external, public and private functions.
+    It is also now impossible to apply modifiers to operators because free functions cannot have them.
+1. Have non-global operators, which has several consequenes:
+    - Operators can only be bound in the file that also defines the type.
+        They are available wherever the type is available and cannot have different definitions in different scopes.
+    - Local bindings at contract level are not possible (i.e. inside contracts and libraries).
+        This also makes it impossible to define operators for local types that are defined inside contracts or libraries.
+1. Define `!`, `<<`, `>>` and `**`.
```
```suggestion
1. Have operators on mixed types.
    Parameter and return value types of operators must be the same.
    Also, no implicit conversions are possible, so it is very simple to determine if a particular operation
    will invoke a user-define operator or a built-in one (just look at the type).
1. Redefine operators.
    You cannot replace any of the built-in operators with a custom definition.
    For user-defined operators, only one definition can be provided.
1. Bind operators to contract functions.
    This also means that operator definitions cannot be inherited, virtual, or remain unimplemented.
1. Bind operators to overloaded functions.
1. Use built-in functions to define operators.
    You cannot e.g. bind an operator directly to `keccak256()`.
1. Change the precedence or associativity of an operator.
1. Define any of the operators with more complex semantics.
    You cannot define:
    - `&&` and `||` due to their short-circuiting behavior.
    - `++` and `--` because they modify their argument and also have separate prefix and postfix variants with different semantics.
    - `+=`, `-=`, `^=`, `*=`, `/=`, `%=`, `&=`, `\|=`, `^=`, `<<=`, `>>=` because they modify their argument.
    - `=`, `()`, `?:` and `.` because they are built-in for all types.
    - `delete`, `new`, and `[]` because they are not usable with UDVTs.
1. Define custom operators that do not exist in the language, like `===` or unary `+`.
1. Bind operators in bulk or accidentally.
    The `using L for T` syntax for attaching a whole library to a type does not cover operators.
    Operators must always be bound explicitly.

In addition to these restrictions, we decided to limit them even more in response to feedback we received before releasing Solidity 0.8.18.
This means you cannot also:
1. Use library functions to define operators.
    This excludes external, public, and private functions.
    It is also now impossible to apply modifiers to operators because free functions cannot have them.
1. Have non-global operators, which has several consequences:
    - Operators can only be bound in the file that also defines the type.
        They are available wherever the type is available and cannot have different definitions in different scopes.
    - Local bindings at the contract level are impossible (i.e. inside contracts and libraries).
        This also makes it impossible to define operators for local types that are defined inside contracts or libraries.
1. Define `!`, `<<`, `>>` and `**`.
```

cameel replied:
> the contract level

https://github.com/ethereum/solidity-blog/pull/202#discussion_r1113401728
> only allowed at file level

I'm getting contradictory suggestions here :) Which one is it?

`at contract level`/`at file level` sounds more correct to me because I'm not referring to any particular file or contract. I'm basically using it as an adjective.

matheusaaguiar reviewed (commented):
I added a more complex example following cameel's suggestions. I am not sure if this makes sense, because I am still a n00b regarding solidity and blockchain in general :sweat_smile: . 
Please, correct my mistakes and point what should I do more or differently.

cameel commented on _posts/2023-02-22-user-defined-operators.md:
````diff
@@ -264,6 +264,130 @@ contract C {
 
+##### Token Conversion
+```solidity
+// ==== File: TokenOperators.sol ====
````
I usually put each file in a separate code block, with name given above the block.

cameel commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,484 @@
+of a bigger expression, even if its type cannot be named or even assigned to anything.
+
+#### `using for`
+Since operators are a purely syntactic element, you can always use ordinary functions as a substitute.
```
> I would have thought the ordinary functions are free functions without "using for". 

Well, it's any function. I mean, you can just call a function anywhere where you'd use an operator and often you have many kinds of functions to choose from. Changed to just `a function`.

> Maybe we should not compare the existing "using for" with operators but rather explain it in isolation?

I'd prefer to keep it that way, because really the main reason I added it is because of operators and I wanted to underscore the elements that make them similar in some ways.

cameel commented on _posts/2023-02-22-user-defined-operators.md:
````diff
@@ -264,6 +264,130 @@ contract C {
 }
 ```
 
+##### Token Conversion
````
I think it's good enough.

cameel commented:
@matheusaaguiar I adjusted your example to properly separate definitions from `using for` and to add binary/unary minus on the same type. Also, did some renaming and structural changes.

Overall, the example is probably good enough, though unfortunately still more contrived than I was hoping for (what do you think @chriseth?). There are some things that maybe we could still improve. I'll point them out in comments below.

cameel commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,606 @@
+
+library A {
+    function toBase(TokenA x) internal pure returns (BaseToken) {
+        return BaseToken.wrap(TokenA.unwrap(x) / int(10 ** TOKEN_A_DIGITS));
```
@matheusaaguiar
This conversion loses all fractional digits. It reduces the number to the integer part. I think it would be better to instead scale it up, to something like `max(TOKEN_A_DIGITS, TOKEN_B_DIGITS)`.

This does not affect the operators, but still looks weird.

matheusaaguiar replied:
Yeah, I was lacking a better idea and just wrote this... I am going to implement your suggestion.

matheusaaguiar replied:
I changed it so that when converting to `BaseToken` the value is upscaled and the reverse happens when converting back to either `A` or `B`.

cameel commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,606 @@
+        address owner = msg.sender;
+        require(balanceOfTokenA[owner] >= amount);
+
+        balanceOfTokenA[owner] = balanceOfTokenA[owner] + -amount;
```
@matheusaaguiar
This is a quick and dirty way I used to shoehorn the unary `-` into the example. Maybe we could come up with something better though.

matheusaaguiar replied:
I though about that, but does it make sense to have `-`, since the underlying type is `uint`?

cameel commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,606 @@
+    function deposit(address receiver, TokenA amount) public {
+        balanceOfTokenA[receiver] = balanceOfTokenA[receiver] + amount;
+        BaseToken depositSize = amount.toBase();
+        averageDepositSize = (numberOfDeposits * averageDepositSize + depositSize) / (numberOfDeposits + ONE);
```
The bad thing about this is that UDVTs should allow us to avoid using the same type for a plain number and a deposit size. I don't see who we can do better though. We can't have operators on mixed types.

matheusaaguiar replied:
Yeah, I felt that need (operators on mixed types) here.

cameel replied:
One thing you could do would be to use attached functions in places where you would use such operators. We should actually show that off for `exp()`.

matheusaaguiar replied:
I replaced the `+ ONE` operation with an attached function `increment`. 
Not sure how we would show that for `**`... or do you mean, we should come up with some function that makes use of `**` for UDVTs?

chriseth commented on _posts/2023-02-22-user-defined-operators.md:
````diff
@@ -0,0 +1,606 @@
+```solidity
+pragma solidity ^0.8.19;
+
+import "./TokenOperators.sol";
````
Should we use explicit imports? This way we needlessly clutter the namespace. We could also use `import "./TokenOperators.sol" as op;` I think this syntax also makes it much clearer where identifiers come from.

cameel replied:
Good point. I did not do it to avoid having to import it all manually and making the example even longer but the `import ... as op` syntax seems to be a good compromise. Changed.

cameel commented:
On @ekpyron's request I removed the speculative last section on the future of operators.

matheusaaguiar commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,606 @@
+using {B.toBase, B.toA} for TokenB;
+
+function addA(TokenA x, TokenA y) pure returns (TokenA) {
+    return (x.toBase() + y.toBase()).toA();
```
Maybe if `A` and `B` value is computed according to a variable rate depending on `Base`?
I have no clue if this makes sense...

franzihei commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,598 @@
+category: Explainers
+---
+
+The prominent new feature in [Solidity 0.8.19](https://github.com/ethereum/solidity/releases/tag/v0.8.19)
```
```suggestion
The most notable new feature in [Solidity 0.8.19](https://github.com/ethereum/solidity/releases/tag/v0.8.19)
```
might sound better, but just a recommendation

NunoFilipeSantos replied:
I agree that prominent does not do it justice, but @franzihei 's alternative made it too wordy.
I would say this feature is a highlight of the release by calling it "The outstanding feature in..."

cameel replied:
`most notable` does sound better to me. `outstanding` not so much.

But `highlight` actually sounds even better. How about `The highlight of Solidity 0.8.19 is ...`?

franzihei commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,598 @@
+---
+
+The prominent new feature in [Solidity 0.8.19](https://github.com/ethereum/solidity/releases/tag/v0.8.19)
+is the support for defining *operators* on [user-defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs).
```
```suggestion
is support for defining *operators* on [user-defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs).
```

NunoFilipeSantos replied:
The article "the" is correct. 🤔

franzihei replied:
for me it sounded weird but you might be right here 👍

franzihei commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,598 @@
+Before we get into the details of the new feature and how it affects the language, it may be a good
+idea to first go over the two existing features that it builds upon: built-in operators and `using for`.
+
+#### Built-in operators
```
```suggestion
#### Built-in Operators
```

Keeping capitalisation consistent in all titles

franzihei commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,598 @@
+| Composite assignment | `+=`, `-=`, `^=`, `*=`, `/=`, `%=`, `&=`, `\|=`, `^=`, `<<=`, `>>=`      |
+| Miscellaneous        | `delete`, `new`, `.`, `()`, `[]`, `?:`                                   |
+
+It is important to note that in many languages, an assignment is a statement that is handled specially.
```
```suggestion
It is important to note that in many languages, an assignment is a statement that is handled uniquely.
```

"Specially" sounds odd to me - maybe rather ``uniquely`` or ``distinctively``?

cameel replied:
Odd, it sounded fine to me. I think I see it quite often. It's `uniquely` and `distinctively` that seem off to me in this context :)

But `in a special way` is good. I will change to that.

franzihei commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,598 @@
+- ✅ At file level.
+- ✅ Inside contract and library definitions.
+
+Can not be used:
```
```suggestion
The directive can not be used:
```

franzihei commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,598 @@
+    operator and specifies the ID of the function used to define it.
+
+This effectively introduces a new way to perform a function call not represented by a `FunctionCall` node.
+This will affect any tool that needs to detect and follow function calls, for example, to build a
```
```suggestion
It will affect any tool that needs to detect and follow function calls, for example, to build a
```
repetition of "this"

franzihei commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,598 @@
+    Operators must always be bound explicitly.
+
+In addition to these restrictions, we decided to limit them even more in response to feedback we received before releasing Solidity 0.8.18.
+This means you cannot also:
```
```suggestion
This means you also cannot:
```

franzihei reviewed (commented):
Added a few small remarks. 

Maybe also add one last sentence to add feature feedback on this to the forum with a link?

cameel commented:
@franzihei
> Maybe also add one last sentence to add feature feedback on this to the forum with a link?

You mean, link specifically to the feature preview thread or just encourage people to post feedback on the forum?

ekpyron commented on _posts/2023-02-22-user-defined-operators.md:
```diff
@@ -0,0 +1,660 @@
+To avoid having to introduce breaking changes later, we decided to defer support for such operators until
+these changes are behind us.
+
+#### Operators on Mixed Types
+Inability to apply operators to values of different types is somewhat constraining and lifting this
+limitation is something we would like to address eventually, though it is not yet clear in what form.
+There are multiple ways to achieve this.
+Explicitly defining an operator for every type combination is just one of them, and not necessarily the most elegant.
+For the time being, we can only say that the feature we are releasing is a minimal viable implementation
+that will definitely not cause problems if considered as part of a generic type system or trait system
+with type inference.
```
This still has to be dropped.

cameel replied:
Dropped.

ekpyron reviewed (approved):
For the record, I don't like the phrasing and terminology - it's not actually user-defined operators - it's implementations for builtin operators on user-defined types. Operators aren't "bound", but you specialize them for types, implement them for types (or really instantiate types to type classes defining them, but that we can only start saying once we have type classes)... but since people need to sleep, I'll approve it anyways :-).
